scisim/static/js/Engines/Patient/PatientEngine.js

```diff
@@ -1,11 +1,146 @@
+
+		$.extend(pageContext, directiveContext);
+		$.extend(pageContext, renderer.composePage(pageContext));
+
+		if(patient !== null){
+			$.extend(pageContext, {"patient": patient});
+		}
+
+		chain.add(pageContext);
+
+		that.applyListeners();
 	});
-};
\ No newline at end of file
+};
+
+PatientEngine.prototype.restorePage = function(page_id){
+	var context = chain.findById(page_id);
+	var patient = context.patient || false;
+
+	new PatientPageDirectiveApplicator(context.page_actions, context.page_modifiers, patient);
+
+	$.extend(context, renderer.composePage(context));
+
+	this.applyListeners();
+};
+
+PatientEngine.prototype.applyListeners = function () {
+	$('#continue-btn').click(this.onContinueClick.bind(this));
+	$('#submit-btn').click(this.onSubmitButtonClick.bind(this));
+	$('.choice-binary .well').click(this.onBinaryChoiceClick.bind(this));
+};
+
+PatientEngine.prototype.onContinueClick = function(){
+	this.changePage(chain.getActivePage().goes_to_page);
+};
+
+PatientEngine.prototype.onSubmitButtonClick = function(e){
+	e.stopImmediatePropagation();
+	var $inputs = $('input'),
+		context = chain.getActivePage(),
+		destination = "";
+
+	for (var i = 0; i < $inputs.length; i++) {
+		var input = $inputs.eq(i);
+		var choiceInfo = {};
+
+		if(input.val().length === 0){
+			// TODO: extract this to a form helper class
+			input.addClass("has-danger");
+			input.attr('placeholder', "This field must be filled.");
+			return;
+		}
+
+		destination = context.goes_to_page || false;
+		if(!destination) destination = input.data("destination");
+
+		choiceInfo.choice = input.val();
+		choiceInfo.choice_id = input.data("choice-id");
+		choiceInfo.page_context = context;
+		choiceInfo.prev = input.prev();
+
+		this.choiceLogger.logChoice(choiceInfo);
+	}
+
+	this.changePage(destination);
+};
+
+
+PatientEngine.prototype.onBinaryChoiceClick = function(e){
+	e.stopImmediatePropagation();
+	var $elem = $(e.currentTarget);
+	var context = chain.getActivePage();
+	if($elem.parent().hasClass("disabled")) return;
+
+	var value = $elem.text(),
+		choiceId = $elem.data('choice-id'),
+		destinationId = $elem.data('destination'),
+		choiceInfo = {
+			choice: value,
+			choice_id: choiceId,
+			page_context: context,
+			prev: $('.page-section').last()
+		};
+
+	this.choiceLogger.logChoice(choiceInfo);
+
+	storageHelper.appendJsonArray("choices_made", choiceId);
+
+	$elem.parent().addClass("disabled");
+
+	this.changePage(destinationId);
+};
+
+PatientEngine.prototype.restoreLast = function(){
+	var context = chain.getLastPage();
+	this.changePage(context.id);
+};
+
+PatientEngine.prototype.disableChoices = function() {
+	var choices = $('.choice');
+	for (var i = 0; i < choices.length; i++) {
+		choices.eq(i).addClass("disabled");
+	}
+};
+
+PatientEngine.prototype.changePage = function(destination) {
+	this.choiceLogger.flushLog();
```

Comment:
it doesn't look like the new page will be rendered before the api requests finish, however i don't think this is a problem. If it does become one, we can wrap it in a deferred.